Surrogates API Documentation

bpaul4 opened the pull request:
## Fixes


## Summary/Motivation:


## Changes proposed in this PR:
- New documentation for surrogates API (ALAMO, PySMO, Keras) Python wrappers
- This new documentation will be the default, and the old ALAMO/PySMO documentation will still exist (for now)

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

bpaul4 commented:
I decided to keep the old ALAMO documentation for now, since some more advanced capabilities such as custom constraints and confidence intervals are not yet supported in the new API. @andrewlee94 if you have the time, I would appreciate your thoughts/feedback on this first pass at the ALAMO documentation and what to do about the old documentation and modules.

andrewlee94 commented:
@bpaul4 I haven't had a chance to look at the new stuff yet, but get rid of the old ALAMO docs - they are no longer supported, so even if they have additional features they aren't of any use.

andrewlee94 commented on docs/explanations/modeling_extensions/surrogate/alamopy/alamopy-cli.rst:
```diff
@@ -14,32 +14,57 @@ Basic ALAMOPY.ALAMO options
 Data Arguments
```
Is this actually correct? I believe the command line interface only applies to the old ALAMO wrapper (and this doc is explicitly discussing a cli). If that is true, then we should just remove this documentation as it is no longer supported (also, the changes here would not be correct as the old API would still apply).

andrewlee94 commented on docs/explanations/modeling_extensions/surrogate/alamopy/alamopy-depr.rst:
```diff
@@ -0,0 +1,167 @@
+.. alamopy documentation master file, created by
```
Drop all of this - we do not want to advertise the old interface, as it is only there to retain internal functionality for RIPE and HELMET.

andrewlee94 reviewed (changes requested):
I just had a quick look through, and we should get rid of most of this - it all applies to the old ALAMO interface which we do not want to advertise anymore.

We should write a completely new set of docs for the new API, with subsections for each of the tools we have available (ALAMO, OMLT, PySMO, etc.).

bpaul4 commented:
@carldlaird @OOAmusat would you be able to provide content for the OMLT/Keras and PySMO index pages? If you have time, I would appreciate your thoughts on the new doctree structure, as well.

bpaul4 commented on docs/explanations/modeling_extensions/surrogate/api/pysmo/pysmo_polyregression.rst:
```diff
@@ -21,111 +22,161 @@ Thus, for a problem with :math:`m` sample points and :math:`n` input variables,
 ------------
-To generate a polynomial model with PySMO, the  *pysmo.polynomial_regression* class is first initialized,
-and then the method ``training`` is called on the initialized object:
+To generate a polynomial model with PySMO, the  *pysmo_surrogate.PysmoPolyTrainer* trainer is instantiated and initialized with the desired configuration arguments, and the the training function ``train_surrogate`` is called:
```
```suggestion
To generate a polynomial model with PySMO, the  *pysmo_surrogate.PysmoPolyTrainer* trainer is instantiated and initialized with the desired configuration arguments, and the training function ``train_surrogate`` is called:
```

bpaul4 commented on docs/explanations/modeling_extensions/surrogate/api/pysmo/pysmo_polyregression.rst:
```diff
@@ -21,111 +22,161 @@ Thus, for a problem with :math:`m` sample points and :math:`n` input variables,
-object simply passing a list of variables into the function *generate_expression*:
+Confidence intervals for *PysmoPolyTrainer* models
+--------------------------------------------------------------------
+**[Needs to be moved over to new interface ---coming soon]**
```
```suggestion
**Coming soon**
```

bpaul4 commented on docs/explanations/modeling_extensions/surrogate/api/pysmo/pysmo_radialbasisfunctions.rst:
```diff
@@ -42,83 +46,128 @@ PySMO offers a range of basis function transformations :math:`\psi`, as shown in
 ------------
-To generate an RBF model with PySMO, the  *pysmo.radial_basis_function* class is first initialized,
-and then the function *training* is called on the initialized object:
+To generate an RBF model with PySMO, the  *pysmo_surrogate.PysmoRBFTrainer* trainer is instantiated and initialized with the desired configuration arguments, and the the training function ``train_surrogate`` is called:
```
```suggestion
To generate an RBF model with PySMO, the  *pysmo_surrogate.PysmoRBFTrainer* trainer is instantiated and initialized with the desired configuration arguments, and the training function ``train_surrogate`` is called:
```

bpaul4 commented on docs/explanations/modeling_extensions/surrogate/sampling/pysmo_sampling_properties.rst:
```diff
@@ -39,11 +46,36 @@ The following code snippet shows basic usage of the package for selecting sample
    >>> samples = space_init.sample_points()
 
+.. note::
+  In the above case, it is assumed that there is only one output variable AND it is in the last column of the dataset. When there are multiple output columns, the xlabels and ylabels options must be specifiec (see below example).
```
```suggestion
  In the above case, it is assumed that there is only one output variable AND it is in the last column of the dataset. When there are multiple output columns, the xlabels and ylabels options must be specified (see below example).
```

andrewlee94 commented on docs/explanations/modeling_extensions/surrogate/api/alamopy/alamopy-options.rst:
```diff
@@ -0,0 +1,196 @@
+
+.. contents::
+    :depth: 3
+
```
The first thing we need here is instructions on how to install ALAMO, emphasizing that it is an external package that needs to be installed manually. We should also mention the need for a license. Most of this can be done with a link to the ALAMO docs. See https://github.com/IDAES/idaes-pse/issues/85.

We should also mention that more detail on ALAMO options can be found in the ALAMO docs.

andrewlee94 commented on docs/explanations/modeling_extensions/index.rst:
```diff
@@ -17,42 +17,55 @@ provided below.
 
-:ref:`ALAMOPY<explanations/modeling_extensions/surrogate/alamopy/index:ALAMOPY: ALAMO Python>`
+:ref:`ALAMOPY: ALAMO Python<explanations/modeling_extensions/surrogate/api/alamopy/index:ALAMOPY: ALAMO Python>`
 provides a wrapper for the software ALAMO which generates algebraic surrogate models of 
```
We should probably add a note block here to say that ALAMO is an external tool that needs to be installed separately as well.

andrewlee94 commented on docs/explanations/modeling_extensions/index.rst:
```diff
@@ -17,42 +17,55 @@ provided below.
+.. rubric:: OMLT: Optimization and Machine Learning Toolkit
+
+:ref:`OMLT: Optimization and Machine Learning Toolkit<explanations/modeling_extensions/surrogate/api/omlt-keras/index:OMLT: Optimization and Machine Learning Toolkit>`
+provides tools for generating machine learning, neural network and gradient-boosted tree models with the Pyomo optimization environment. It provides IDAES users optimization formulations for ML models (e.g. full-space, reduced-space, MILP) and an interface to import Sequential Keras and general ONNX models.
```
Similar to the other comments, we should mention that this requires additional dependencies - details can be left to the more in depth docs page.

andrewlee94 reviewed (commented):
@bpaul4 What is the status of this PR? If all we are waiting on is the OMLT/Keras docs, then we should just move ahead with a placeholder there for now and create a new issue for @carldlaird  to address later. We should get what we have out for the November release, as this has lingered too long.

I just added a few minor comments that come from an old issue that we should fix here.

bpaul4 commented:
@andrewlee94 once I've addressed your comments above (will do that today), the landing page and ALAMO/PySMO sections will be complete. I'll clean up the OMLT placeholder page so it looks presentable in lieu of detailed content.

andrewlee94 reviewed (approved):
Thanks for your effort to put this together @bpaul4 .

bpaul4 commented:
Thank you @OOAmusat and @andrewlee94 for your contributions. I'll continue to coordinate with @carldlaird to complete the OMLT pages for a future update.

andrewlee94 commented:
@bpaul4 Could you please create a new issue for the OMLT docs, and assign @carldlaird to it (and yourself if appropriate).

OOAmusat reviewed (approved):
LGTM
